path_subkeys.go

```diff
@@ -0,0 +1,190 @@
+		Fields: map[string]*framework.FieldSchema{
+			"path": {
+				Type:        framework.TypeString,
+				Description: "Location of the secret.",
+			},
+			"depth": {
+				Type:        framework.TypeInt,
+				Description: "The maximum depth to traverse. No limit will be imposed if not provided or if 0.",
+			},
+			"version": {
+				Type:        framework.TypeInt,
+				Description: "Specifies which version to retrieve. If not provided, the current version will be used.",
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation: b.upgradeCheck(b.pathSubkeysRead()),
+		},
+
+		HelpSynopsis:    subkeysHelpSyn,
+		HelpDescription: subkeysHelpDesc,
+	}
+}
+
+// removeValues recursively walks the provided secret data represented as a
+// map. All leaf nodes (i.e. empty maps and non-map values) will be replaced
+// with nil in an effort to remove all values. The resulting structure will
+// provide all subkeys with nesting fully intact. The modifications are made
+// to the input in-place. maxDepth will denote how deep to traverse. A maxDepth
+// of 0 is the equivalent of no limit.
+func removeValues(input map[string]interface{}, maxDepth int) {
+	var walk func(interface{}, int)
+
+	walk = func(in interface{}, depth int) {
+		val := reflect.ValueOf(in)
+
+		if val.Kind() == reflect.Map {
+			for _, k := range val.MapKeys() {
+				v := val.MapIndex(k)
+				m := in.(map[string]interface{})
+
+				switch t := v.Interface().(type) {
+				case map[string]interface{}:
+					// Only continue walking if we have not reached max depth
+					// and the underlying map has at least 1 key. The key is
+					// otherwise treated as a leaf node and thus set to nil.
+					// Setting to nil if the max depth is reached is crucial in
+					// that it prevents leaking secret data as the input map is
+					// being modified in-place
+					if currentDepth := depth + 1; (maxDepth == 0 || currentDepth <= maxDepth) && len(t) > 0 {
+						walk(t, currentDepth)
+					} else {
+						m[k.String()] = nil
+					}
+				default:
+					m[k.String()] = nil
+				}
+			}
+		}
+	}
+
+	walk(input, 1)
+}
+
+// pathSubkeysRead handles ReadOperation requests for a specified path. Subkeys
+// that exist within the entry specified by the provided path will be retrieved.
+// This is done by stripping the secret data by replacing all underlying values of
+// leaf keys with null.
+func (b *versionedKVBackend) pathSubkeysRead() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		key := data.Get("path").(string)
+		depth := data.Get("depth").(int)
+
+		lock := locksutil.LockForKey(b.locks, key)
+		lock.RLock()
+		defer lock.RUnlock()
+
+		meta, err := b.getKeyMetadata(ctx, req.Storage, key)
+		if err != nil {
+			return nil, err
+		}
+		if meta == nil {
+			return nil, nil
+		}
+
+		versionNum := meta.CurrentVersion
+		versionParam := data.Get("version").(int)
```

Comment:
I could be wrong on this, but if memory serves we need to do `GetOk` for parameters that are not required?